CRM/Core/BAO/MessageTemplate.php

```diff
@@ -692,35 +694,26 @@ protected static function parseThroughSmarty(array $mailContent, $tplParams): ar
    *
    * @param array $mailContent
    * @param bool $disableSmarty
-   * @param int $contactID
+   * @param int|NULL $contactID
    * @param array $smartyAssigns
    *
    * @return array
-   * @throws \CRM_Core_Exception
    */
-  public static function renderMessageTemplate(array $mailContent, $disableSmarty, $contactID, $smartyAssigns): array {
-    $tokens = self::getTokensToResolve($mailContent);
-
-    // When using Smarty we need to pass the $escapeSmarty parameter.
-    $escapeSmarty = !$disableSmarty;
-
-    $mailContent = self::resolveDomainTokens($mailContent, $tokens, $escapeSmarty);
-
-    if ($contactID) {
-      $mailContent = self::resolveContactTokens($contactID, $tokens, $mailContent, $escapeSmarty);
-    }
-
-    // Normally Smarty is run, but it can be disabled using the disableSmarty
-    // parameter, which may be useful for non-core uses of MessageTemplate.send
-    // In particular it helps with the mosaicomsgtpl extension.
-    if (!$disableSmarty) {
-      $mailContent = self::parseThroughSmarty($mailContent, $smartyAssigns);
-    }
-    else {
-      // Since we're not relying on Smarty for this function, we DIY.
-      // strip whitespace from ends and turn into a single line
-      $mailContent['subject'] = trim(preg_replace('/[\r\n]+/', ' ', $mailContent['subject']));
+  public static function renderMessageTemplate(array $mailContent, bool $disableSmarty, $contactID, array $smartyAssigns): array {
+    CRM_Core_Smarty::singleton()->pushScope($smartyAssigns);
+    $tokenProcessor = new TokenProcessor(\Civi::dispatcher(), ['smarty' => !$disableSmarty]);
+    $tokenProcessor->addMessage('html', $mailContent['html'], 'text/html');
+    $tokenProcessor->addMessage('text', $mailContent['text'], 'text/plain');
+    $tokenProcessor->addMessage('subject', $mailContent['subject'], 'text/plain');
+    $tokenProcessor->addRow($contactID ? ['contactId' => $contactID] : []);
```

Comment:
@totten @eileenmcnaughton To support rendering other tokens here we just need to pass the ID through here. Eg.
```
[
  'contactId' => $contactID,
  'contributionId' => $contributionID,
]
```
Then it would work with this https://github.com/civicrm/civicrm-core/pull/19584 and contribution tokens would be rendered.
There's a few steps to get there but it's not a long way off - probably review/merge this, review/merge #19550, add testing on contribution tokens like @eileenmcnaughton did with contact tokens, review/merge #16983 + #19071, modify `renderMessageTemplate()` to accept different entity IDs and get that passed through from eg. contribution invoice workflow template.

Reply:
@mattwire - I would definitely want to include full test cover of the supported contribution tokens with any changes.

Also I'd want to delay exposing the api (incl sendtemplate api) until it's settled